pkg/handlers/apk.go

```diff
@@ -0,0 +1,403 @@
+func (h *apkHandler) handleAPKFileContent(ctx logContext.Context, rdr io.Reader, fileName string, apkChan chan []byte) error {
+	mimeReader, err := newMimeTypeReader(rdr)
+	if err != nil {
+		return fmt.Errorf("failed to create mimeTypeReader for file %s: %w", fileName, err)
+	}
+	ctx = logContext.WithValues(
+		ctx,
+		"filename", fileName,
+	)
+	return h.handleNonArchiveContent(ctx, mimeReader, apkChan)
+}
+
+// createZipReader creates a new ZIP reader from the input fileReader.
+func createZipReader(input fileReader) (*zip.Reader, error) {
+	size, err := input.Size()
+	if err != nil {
+		return nil, err
+	}
+	zipReader, err := zip.NewReader(input, size)
+	if err != nil {
+		return nil, err
+	}
+	return zipReader, err
+}
+
+// parseResTable parses the resources.arsc file and returns the ResourceTable.
+func parseResTable(zipReader *zip.Reader) (*apkparser.ResourceTable, error) {
+	for _, file := range zipReader.File {
+		if file.Name == "resources.arsc" {
+			rdr, err := openFile(file)
+			if err != nil {
+				return nil, err
+			}
+
+			resTable, err := apkparser.ParseResourceTable(rdr)
+			rdr.Close()
+			if err != nil {
+				return nil, err
+			}
+			return resTable, nil
+		}
+	}
+	return nil, errors.New("resources.arsc file not found in the APK archive")
+}
+
+// openFile opens the file from the zip archive and returns the data as an io.ReadCloser
+// Note: responsibility of calling function to close the reader
+func openFile(file *zip.File) (io.ReadCloser, error) {
+	rc, err := file.Open()
+	if err != nil {
+		return nil, err
+	}
+	return rc, nil
+}
+
+// extractStringsFromResTable extracts the strings from the resources table
+// Note: This is a hacky way to get the strings from the resources table
+// APK strings are typically (always?) stored in the 0x7f000000-0x7fffffff range
+// https://chromium.googlesource.com/chromium/src/+/master/build/android/docs/life_of_a_resource.md
+func extractStringsFromResTable(resTable *apkparser.ResourceTable) (io.Reader, error) {
+	var resourceStrings bytes.Buffer
+	inStrings := false
+	for i := 0x7f000000; i <= 0x7fffffff; i++ {
+		entry, _ := resTable.GetResourceEntry(uint32(i))
+		if entry == nil {
+			continue
+		}
+		if entry.ResourceType == "string" {
+			inStrings = true
+			val, err := entry.GetValue().String()
+			if err != nil {
+				return nil, err
+			}
+			// Write directly to the buffer
+			resourceStrings.WriteString(entry.Key)
+			resourceStrings.WriteString(": ")
+			resourceStrings.WriteString(val)
+			resourceStrings.WriteString("\n")
+		}
+		// Exit the loop if we've finished processing the strings
+		if inStrings && entry.ResourceType != "string" {
+			break
+		}
+	}
+	return &resourceStrings, nil
+}
+
+// processDexFile decodes the dex file and returns the relevant instructions
+func processDexFile(ctx logContext.Context, rdr io.ReadCloser) (io.Reader, error) {
+	// dextk.Read() requires an io.ReaderAt interface
+	dexReader, err := dextk.Read(iobuf.NewBufferedReaderSeeker(rdr))
```

Comment:
Would this implementation still provide perf improvements? `dexReader, err := dextk.Read(iobuf.NewBufferedReaderSeeker(rdr), dextk.WithReadCache(16))` You can't pass in the `rdr` object as is b/c it needs an `io.ReaderAt`.

Reply:
BufferedReadSeeker wraps the rdr and implements the io.ReaderAt interface. I also tested it with some benchmarks to confirm the performance gain. I can share them here when I'm back at my computer.

Reply:
Consider constructing the BufferedReadSeeker at the call site and passing it into this function. Changing the function’s argument to a fileReader would allow you to use rdr directly.

Reply:
I wasn’t sure what value to use for the number of slots. I chose 8 since each slot is 4KB, making 32KB feel reasonable, but I’m open to increasing it. What value were you testing with? I think 16 could also work, but I’ll leave this decision to you.

It would’ve been helpful if the library had code comments or documentation to guide us. I was just exploring the code to understand how that cache is used. 😅